elide-core/src/main/java/com/yahoo/elide/jsonapi/extensions/JsonApiAtomicOperations.java

```diff
@@ -0,0 +1,563 @@
+            actions = operations.getOperations();
+        } catch (InvalidFormatException e) {
+            if (e.getMessage() != null
+                    && e.getMessage().contains("$OperationCode")) {
+                // Invalid op code results in a format error as it is an enum
+                throw new InvalidEntityBodyException(
+                        "Invalid Atomic Operations extension operation code:"
+                                + e.getValue());
+            } else {
+                throw new InvalidEntityBodyException(operationsDoc);
+            }
+        } catch (IOException e) {
+            throw new InvalidEntityBodyException(operationsDoc);
+        }
+        JsonApiAtomicOperations processor = new JsonApiAtomicOperations(dataStore, actions, uri, requestScope);
+        return processor.processActions(requestScope);
+    }
+
+    /**
+     * Constructor.
+     *
+     * @param dataStore Data Store
+     * @param actions List of patch actions
+     * @param rootUri root URI
+     */
+    private JsonApiAtomicOperations(DataStore dataStore,
+            List<Operation> actions,
+            String rootUri,
+            RequestScope requestScope) {
+        this.actions = actions.stream().map(OperationAction::new).toList();
+        this.rootUri = rootUri;
+    }
+
+    /**
+     * Process atomic operations actions.
+     *
+     * @return Pair (return code, JsonNode)
+     */
+    private Supplier<Pair<Integer, JsonNode>> processActions(JsonApiAtomicOperationsRequestScope requestScope) {
+        try {
+            List<Supplier<Pair<Integer, JsonApiDocument>>> results = handleActions(requestScope);
+
+            postProcessRelationships(requestScope);
+
+            return () -> {
+                try {
+                    return Pair.of(HttpStatus.SC_OK,
+                            mergeResponse(results, requestScope.getMapper()));
+                } catch (HttpStatusException e) {
+                    throwErrorResponse();
+                    // NOTE: This should never be called. throwErrorResponse should _always_ throw an exception
+                    return null;
+                }
+            };
+        } catch (HttpStatusException e) {
+            throwErrorResponse();
+            // NOTE: This should never be called. throwErrorResponse should _always_ throw an exception
+            return () -> null;
+        }
+    }
+
+    protected String getFullPath(Ref ref, Operation operation) {
+        if (ref != null) {
+            StringBuilder fullPathBuilder = new StringBuilder();
+            if (ref.getType() == null) {
+                throw new InvalidEntityBodyException(
+                        "Atomic Operations extension requires ref type to be specified.");
+            }
+            fullPathBuilder.append(ref.getType());
+
+            // Only relationship operations or resource update operation should have the id
```

Comment:
Is this true?  What about remove?  That looks like it can also have an ID.

Reply:
For some reason I recall that the delete needed to be performed on the collection endpoint and needed to have data specified with the type and id. So later on if the data was omitted I will attempt to generate this when specified from the ref.

```java
                    // If data is not provided, but it is to remove a resource the data is optional
                    // and should be generated from the ref
                    if ((data == null || JsonNodeType.NULL.equals(data.getNodeType()))
                            && OperationCode.REMOVE.equals(operation.getOperationCode())) {
                        data = requestScope.getMapper().getObjectMapper().createObjectNode().put("type", ref.getType())
                                .put("id", ref.getId() != null ? ref.getId() : ref.getLid());
                    }
```

Reply:
The JSON API core spec has this example of a DELETE with the ID in the path:
```
DELETE /photos/1 HTTP/1.1
Accept: application/vnd.api+json
```

The operations spec has this example of specifying the ID in the operation:

```json
{
  "atomic:operations": [{
    "op": "remove",
    "ref": {
      "type": "articles",
      "id": "13"
    }
  }]
}
```

handleDeleteOp does not append the ID to the fullPath if data is missing.  However, if data is present it appends the ID.  Not sure what is more common, but the path will be different depending on the presence of data.  The spec example shows data missing.  This looks to me like it might not work.

Reply:
You're correct I'm not sure why I ended up not specifying the id in the ref for the remove operation. I shall change it to use the id in the remove operation and not generate the data.